spec/controllers/users/sessions_controller_spec.rb

```diff
@@ -116,6 +116,46 @@
       end
     end
 
+    context 'locked out session' do
+      let(:locked_at) { Time.zone.now }
+      let(:user) { create(:user, :fully_registered) }
+      let(:bad_password_window) { IdentityConfig.store.max_bad_passwords_window_in_seconds }
+
+      before do
+        allow(subject).to receive(:session_bad_password_count_max_exceeded?).
+          and_return(true)
+        allow(subject).to receive(:clear_session_bad_password_count_if_window_expired).
+          and_return(nil)
+      end
+
+      it 'renders an error letting user know they are locked out for a period of time' do
+        session[:max_bad_passwords_at] = locked_at
+        post :create, params: { user: { email: user.email.upcase, password: user.password } }
+        current_time = Time.zone.now
+        time_in_hours = distance_of_time_in_words(
+          current_time,
+          (locked_at + bad_password_window.seconds),
+          true,
+        )
+
+        expect(response).to redirect_to root_url
+        expect(flash[:error]).to eq(
+          t(
+            'errors.sign_in.bad_password_limit',
+            time_left: time_in_hours,
+          ),
+        )
+      end
+
+      it 'tracks unsuccessful authentication for too many auth failures' do
+        session[:max_bad_passwords_at] = locked_at
+        allow(subject).to receive(:session_bad_password_count_max_exceeded?).and_return(true)
+        mock_email_parameter = { email: 'bob@example.com' }
+
+        post :create, params: { user: { **mock_email_parameter, password: 'eatCake!' } }
+      end
```

Comment:
Similar to my previous comment, I think stubbing controller methods creates a lot of risk that the real method code may not work as we'd expect it to. I think it's fine to stub something _outside_ the controller that we're accessing, since we should assume there's test coverage for those methods elsewhere. But stubbing the controller's own methods from the controller spec creates gaps in coverage, in my opinion.

In this case, I think it'd be pretty straightforward to simulate the session state which triggers the locked out session.

```suggestion
      before do
        session[:bad_password_count] = IdentityConfig.store.max_bad_passwords + 1
        session[:max_bad_passwords_at] = locked_at.to_i
      end

      it 'renders an error letting user know they are locked out for a period of time' do
        post :create, params: { user: { email: user.email.upcase, password: user.password } }
        current_time = Time.zone.now
        time_in_hours = distance_of_time_in_words(
          current_time,
          (locked_at + bad_password_window.seconds),
          true,
        )

        expect(response).to redirect_to root_url
        expect(flash[:error]).to eq(
          t(
            'errors.sign_in.bad_password_limit',
            time_left: time_in_hours,
          ),
        )
      end

      it 'tracks unsuccessful authentication for too many auth failures' do
        mock_email_parameter = { email: 'bob@example.com' }

        post :create, params: { user: { **mock_email_parameter, password: 'eatCake!' } }
```